[REVIEW] Setting correct is_unique and is_monotonic index attributes

rjzamora opened the pull request:
This is a short-term fix for the [Issue#2150](https://github.com/rapidsai/cudf/issues/2150).  The following index attributes will now be set to `False`:

- `is_monotonic`
- `is_monotonic_increasing`
- `is_monotonic_decreasing`
- ` is_unique`

@qinxuye @kkraus14 - Please confirm that this is a reasonable fix (while I work on gettting the attributes set correctly)

[**EDIT: This PR has expanded to include proper `is_unique` and `is_monotonic_*` settings.  It also modifies slicing/indexing behavior to be consistent with pandas for monotonic columns/indices.**]

<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

qinxuye commented:
It looks fine to me.

rjzamora commented:
@kkraus @shwina - Sorry for the delay with this (I have had a lot of connectivity issues for the past 24 hours).  I just pushed some changes to set the attributes "correctly."  Please do make style suggestions or point out mistakes or missed cases.

shwina commented on python/cudf/dataframe/index.py:
```diff
@@ -537,6 +561,26 @@ def find_label_range(self, first, last):
+        return len(set(self._values)) == len(self._values)
+
+    @property
+    def is_monotonic_increasing(self):
```
For `DatetimeIndex` specifically (a subclass of `GenericIndex`), can this not be done on the device?

kkraus14 commented on python/cudf/dataframe/index.py:
```diff
@@ -537,6 +561,26 @@ def find_label_range(self, first, last):
 
+    @property
+    def is_unique(self):
+        return len(set(self._values)) == len(self._values)
```
You can't use `set(...)` here as that doesn't know how to operate against gpu values. You'd likely want `.nunique` for a Series and `.unique_count` for a column.

kkraus14 commented on python/cudf/dataframe/index.py:
```diff
@@ -537,6 +561,26 @@ def find_label_range(self, first, last):
+    @property
+    def is_monotonic_increasing(self):
+        for val in zip(self._values[1:], self._values[:-1]):
+            if val[0] < val[1]:
```
Your idea here is great in using slicing to do a stencil comparison, but we need to rework this a bit to have it run nicely on the device. Instead of zipping them and iterating we can just run a binaryop and then use `all()` to ensure it's all true. This would look like:
```
return (self._values.iloc[1:] >= self._values[:-1]).all()
```

kkraus14 replied:
Also, we should memoize these instead of recomputing them every time 😄

rjzamora replied:
I added some manual memoization - Perhaps I can/should just add a `memoize_property` function to `cudf/utils/utils.py` (and just use a `@memoize_property` decorator throughout)?

kkraus14 commented on python/cudf/dataframe/index.py:
```diff
@@ -537,6 +561,26 @@ def find_label_range(self, first, last):
+
+    @property
+    def is_monotonic_decreasing(self):
+        for val in zip(self._values[1:], self._values[:-1]):
```
Same here:
```
return (self._values.iloc[1:] <= self._values[:-1]).all()
```

kkraus14 commented on python/cudf/dataframe/multiindex.py:
```diff
@@ -465,5 +465,29 @@ def from_pandas(cls, multiindex):
 
+    @property
+    def is_unique(self):
+        return self.codes._size == self.codes.drop_duplicates()._size
```
Instead of using `.codes` here we can use `self._source_data` which will be more efficient.

kkraus14 commented on python/cudf/dataframe/multiindex.py:
```diff
@@ -465,5 +465,29 @@ def from_pandas(cls, multiindex):
+            ]:
+                return False
+        else:
+            return True
```
You should be able to use the same logic as above and just loop through the columns and then make sure everything returns True as expected.

rjzamora replied:
I used pandas as a reference for the current logic.  I am not seeing a simple way to loop through columns (and apply per-column `is_monotonic` logic).  Please do let me know if I am missing something obvious (very possible), or if there is a better solution than `np.lexsort`

kkraus14 commented on python/cudf/dataframe/categorical.py:
```diff
@@ -321,6 +321,26 @@ def find_last_value(self, value):
 
+    @property
+    def is_unique(self):
+        return self.unique_count() == len(self)
```
Can this just be moved up to the `column.py` level and inherit from there?

kkraus14 commented on python/cudf/dataframe/categorical.py:
```diff
@@ -321,6 +321,26 @@ def find_last_value(self, value):
+        if not hasattr(self, '_is_monotonic_increasing'):
+            self._is_monotonic_increasing = self._ordered and (
+                self[1:].ordered_compare('ge', self[:-1])
+            ).all()
```
I think you can change this to reuse the numerical implementation:
```
self._is_monotonic_increasing = self._ordered and self.as_numerical.is_monotonic_increasing
```

kkraus14 commented on python/cudf/dataframe/categorical.py:
```diff
@@ -321,6 +321,26 @@ def find_last_value(self, value):
+        if not hasattr(self, '_is_monotonic_decreasing'):
+            self._is_monotonic_decreasing = self._ordered and (
+                self[1:].ordered_compare('le', self[:-1])
+            ).all()
```
```
self._is_monotonic_decreasing = self._ordered and self.as_numerical.is_monotonic_decreasing
```

kkraus14 commented on python/cudf/dataframe/column.py:
```diff
@@ -602,5 +603,29 @@ def _pointer(self):
+        if not hasattr(self, '_is_monotonic_increasing'):
+            self._is_monotonic_increasing = (
+                series.Series(self)[1:] >= series.Series(self)[:-1]
+            ).all()
```
We shouldn't need to construct a Series in order to do this here. You should be able to use the binop functionality to do this.

kkraus14 replied:
This will actually not do what you want as Slicing the Series will misalign the indexes and then the binaryop will realign them as well.

rjzamora replied:
Not sure I understand your second comment.  However, I added `numeric_column_binop`-based methods to `NumericalColumn` (so the unused `Column` methods could be removed) - Is this reasonable?

kkraus14 replied:
calling `series.Series(self)` will construct a Series with a index of 0 --> N (where N is the length of the Series), then slicing it with `1:` will make the make the Index 1 --> N, while slicing it with `:-1` will make the index 0 --> N-1. When doing the `>=` binaryop the first thing it will do is align the two Series based on their indexes which is not what you want to do as everything will always return `True`.

kkraus14 commented on python/cudf/dataframe/column.py:
```diff
@@ -602,5 +603,29 @@ def _pointer(self):
+            self._is_monotonic_decreasing = (
+                series.Series(self)[1:] <= series.Series(self)[:-1]
+            ).all()
+        return self._is_monotonic_decreasing
```
Here too.

kkraus14 commented on python/cudf/dataframe/multiindex.py:
```diff
@@ -465,5 +465,32 @@ def from_pandas(cls, multiindex):
+            values = [
+                self._source_data[i] for i in reversed(self._source_data._cols)
+            ]
+            sort_order = np.lexsort(values)
```
Instead of using `np.lexsort` you should be able to use `cudf.DataFrame.argsort` to do the same as the lexsort here, except it will run on device 😄 and return you a NumericalColumn that you can check the monotonic properties of.
```
self._source_data.argsort().is_monotic_increasing
```

rjzamora replied:
Perfect, that’s exactly what I was looking for - Thanks!

kkraus14 commented on python/cudf/dataframe/multiindex.py:
```diff
@@ -465,5 +465,32 @@ def from_pandas(cls, multiindex):
+    @property
+    def is_monotonic_decreasing(self):
+        if not hasattr(self, '_is_monotonic_decreasing'):
+            self._is_monotonic_decreasing = self[::-1].is_monotonic_increasing
```
```
self._source_data.argsort(ascending=True).is_monotonic_decreasing
```

kkraus14 commented on python/cudf/dataframe/datetime.py:
```diff
@@ -249,6 +249,26 @@ def find_last_value(self, value):
 
+    @property
+    def is_unique(self):
+        return self.as_numerical.unique_count() == len(self)
```
Just to make it more resilient to implementation changes in the future.
```suggestion
        return self.as_numerical.is_unique
```

kkraus14 commented:
@rjzamora @quasiben it looks like in properly calculating the monotonic properties of indexes we're taking a different codepath within dask-cudf that isn't fully supported yet.

kkraus14 commented:
Looks like we just need to implement: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Index.get_slice_bound.html which should correspond to our column `find_first_value` and `find_last_value` APIs.

rjzamora commented:
Thanks @kkraus14 -  I was just about to make a note about the missing`get_slice_bound` method. Thanks for pointing out the relevant API :)

rjzamora commented:
@kkraus14 Sorry for the delay on this - I had a silly bug that was only causing problems in windows CI.

I added the necessary `get_slice_bound` functionality (which required small changes to the existing `find_first_value` and `find_last_value` methods).  Note that when slicing **monotonic** columns/indices in pandas, the specified endpoints are not required to exist in the column/index.  I implemented the necessary changes to match pandas behavior for numerical columns (which is required to pass dask_cudf tests), but I did not implement comprehensive solutions for strings/multi-index/range-index/etc.

I suggest that we raise a new issue for more-rigorous changes to `get_slice_bound` and `find_<first,last>_value` after this PR is merged.  Does that sound reasonable?

kkraus14 commented on python/cudf/cudf/utils/cudautils.py:
```diff
@@ -776,22 +778,53 @@ def gpu_mark_found_float(arr, val, out, not_found):
+def gpu_mark_gt(arr, val, out, not_found):
+    i = cuda.grid(1)
+    if i < arr.size:
+        if check_gt(arr[i], val):
```
Was there a reason you needed to turn this into a separate function instead of just doing:
```
if arr[i] > val:
```

rjzamora replied:
You're right - these separate functions shouldn't be necessary (afaik).  I will remove them.  If it becomes necessary to cover edge cases the way `check_equals_float` does, seperate functions *can* be added in the future.

kkraus14 replied:
`check_equals_float` was special cased to handle `NaN` values which are only problematic when we need `NaN` == `NaN` for certain functions.

kkraus14 commented on python/cudf/cudf/utils/cudautils.py:
```diff
@@ -776,22 +778,53 @@ def gpu_mark_found_float(arr, val, out, not_found):
+def gpu_mark_lt(arr, val, out, not_found):
+    i = cuda.grid(1)
+    if i < arr.size:
+        if check_lt(arr[i], val):
```
Same question here
